doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,297 @@
+We might get a score of 0.3 for this result set (from 0 to 1).
+
+We could then investigate why the results are bad. Perhaps there is a
+best bet. Or popularity might be pushing HMRC docs up. We can fix these
+things, then check the new nDCG score, and we might get 0.7 for this query.
+
+In contrast, with Learning To Rank, search result ranks would be
+determined by a machine learning model without manually changing
+the boost of a field.
+
+With Learning To Rank we would go through the same process of
+rating the search results.
+
+```
+harry potter  0         Pottery barn
+harry potter  3         Harry Potter
+...
+```
+
+However, after doing this, we would train a model with this data
+to rank the results better. The model optimises for nDCG.
+
+We train the model using using a combination of relevancy judgements
+and features: numeric measures which indicate how well we think the
+result matches the query.
+
+The text similarity scores Elasticsearch assigns to fields in the
+document are examples of features. The model tries to find correlations
+between document features and relevancy judgements, which allows it to
+predict how to order documents we don't have relevancy judgements for.
+
+```
+query         rating    document      title_score view_count  recency  pagerank ...
+harry potter  0         Pottery barn  0.2         9876        0.12     9876     ...
+harry potter  3         Harry Potter  0.2         1234        1234     1234     ...
+...
+```
+
+For example, take the SVM dataset below for the query 'harry potter' (qid:1)
+
+```
+0 qid:1 popularity:0.3 title_score:0.2 description_score:0.3 recency:0.9 pagerank:5 ...
+3 qid:1 popularity:0.1 title_score:0.6 description_score:0.7 recency:0.5 pagerank:2 ...
+```
+
+These graded query:document pairs and their features can be used
+to train a model.
+
+The model might learn that title and description scores are slightly
+more useful that popularity and recency, and how much, in certain
+circumstances.
+
+Rather than manually tuning boosts, a model learns from relevancy
+judgements the role that features should play when providing relevant results.
+
+This is a high level overview of what Learning To Rank is for our use case.
+There has been a lot of research in this area, and within the fields of
+information retrieval & machine learning which Learning To Rank intersects.
+There is a lot more out there to learn!
+
+For more details about how we plan to implement Learning To Rank, see
+the implementation section below.
+
+## Motivation
+
+We have decided to try Learning To Rank because it has the potential
+to greatly improve search relevance in a way that is:
+
+- automatable
+- scalable
+- more maintainable
+- more successful
+
+**Automatable** because the model won't require human involvement to
+tune field boosts on the latest user data
+
+**Scalable** because automation; but also because the model should
+provide better results with more data.
```

Comment:
When you say scalable, do you mean in the sense that it removes the need for a human to find-tune everything to improve search? This feels somewhat similar to the previous _automatable_. We could also add here, if it's the case, that we're scalable (perhaps adaptable) to improving search for future versions of documents because training the model determines which of the fields (features) are most useful to improving search.

Reply:
Do we still need to provide some human-annotated relevancy judgements to enable LTR to improve search results, or will we be figuring this out from other data (clickthrough data of the result(s) users choose, for example) ? (this fits better under the _maintainable_ heading)

Reply:
We can use both human-generated judgements and implicit judgements from user behaviour for training the model.

At the moment the judgements generated from user data dwarfs the manually created judgements (millions vs ~2000). Using the user generated data is easier, too, since they don't get stale. If the judgements are good quality then I'm not sure if we will need manual judgements, which makes the approach maintainable, which I suppose is part of it being scalable. 🤷‍♂